rs/linux_kernel_command_line/src/lib.rs

```diff
@@ -0,0 +1,354 @@
+/// Utilities to manipulate a kernel command line reliably.
+use std::error::Error as StdError;
+use std::fmt;
+
+#[derive(Debug)]
+/// A kernel command line with improperly-quoted argument values.
+pub struct ImproperlyQuotedValue {
+    val: String,
+}
+impl StdError for ImproperlyQuotedValue {}
+
+impl fmt::Display for ImproperlyQuotedValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Unclosed quote capturing value {:?} from supplied command line",
+            self.val
+        )
+    }
+}
+
+#[derive(Debug)]
+/// A value unrepresentable as a kernel command line argument value.
+pub struct UnrepresentableValue {
+    val: String,
+}
+
+impl StdError for UnrepresentableValue {}
+
+impl fmt::Display for UnrepresentableValue {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Illegal characters in supplied value {:?}", self.val)
+    }
+}
+
+#[derive(Debug, Default)]
+/// Represents a correctly-parsed kernel command line.
+pub struct KernelCommandLine {
+    tokenized_arguments: Vec<String>,
+}
+
+impl KernelCommandLine {
+    fn format_argument(
+        argument: &str,
+        value: Option<&str>,
+    ) -> Result<String, UnrepresentableValue> {
+        Ok(argument.to_string()
+            + &match value {
+                Some(val) => ("=".to_owned()
+                    + (if val.contains("\"") || val.contains("\n") {
+                        return Err(UnrepresentableValue {
+                            val: val.to_string(),
+                        });
+                    } else if val.contains(" ") {
+                        ("\"".to_owned() + val + "\"").to_string()
+                    } else {
+                        val.to_string()
+                    }
+                    .as_str()))
+                .to_string(),
+                None => "".to_string(),
+            })
+    }
+
+    /// Remove an argument from a kernel command line, however many times it appears.
+    /// Returns the position of the first removed argument.
+    pub fn remove_argument(&mut self, argument: &str) -> Option<usize> {
+        let mut firstpos: Option<usize> = None;
+        self.tokenized_arguments = self
+            .tokenized_arguments
+            .iter()
+            .enumerate()
+            .filter(|(pos, arg)| {
+                let res =
+                    *arg != argument && !arg.starts_with((argument.to_owned() + "=").as_str());
+                if !res && firstpos.is_none() {
+                    firstpos.replace(*pos);
+                }
+                res
+            })
+            .map(|(_, x)| x.clone())
+            .collect::<Vec<String>>();
```

Comment:
optional nit: just `.collect()` should work and is a little more elegant/idiomatic